src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp

```diff
@@ -463,6 +482,50 @@ PRAGMA_DIAG_POP
     }
   }
 
+#ifdef __APPLE__
+  // Execution protection violation
+  //
+  // This should be kept as the last step in the triage.  We don't
+  // have a dedicated trap number for a no-execute fault, so be
+  // conservative and allow other handlers the first shot.
+  if (UnguardOnExecutionViolation > 0 &&
+      (sig == SIGBUS)) {
+    static __thread address last_addr = (address) -1;
+
+    address addr = (address) info->si_addr;
+    address pc = os::Bsd::ucontext_get_pc(uc);
+
+    if (pc != addr && uc->context_esr == 0x9200004F) { //TODO: figure out what this value means
+      // We are faulting trying to write a R-X page
+      pthread_jit_write_protect_np(false);
+
+      log_debug(os)("Writing protection violation "
+                    "at " INTPTR_FORMAT
+                    ", unprotecting", p2i(addr));
+
+      stub = pc;
+
+      last_addr = (address) -1;
+    } else if (pc == addr && uc->context_esr == 0x8200000f) { //TODO: figure out what this value means
+      // We are faulting trying to execute a RW- page
+
+      if (addr != last_addr) {
+        pthread_jit_write_protect_np(true);
+
+        log_debug(os)("Execution protection violation "
+                      "at " INTPTR_FORMAT
+                      ", protecting", p2i(addr));
+
+        stub = pc;
+
+        // Set last_addr so if we fault again at the same address, we don't end
+        // up in an endless loop.
+        last_addr = addr;
+      }
+    }
+  }
+#endif
+
```

Comment:
This looks like you switch rw- <-> r-x state on demand (sigbus) here?
If so, is it good idea from security point ?
This protection is supposed to "protect" us from executing unsigned code ( in map_jit) that are not generated by our jit compiler. 
For example if executable code (in map_jit area) is modified by another executable code in (map_jit area).

This code makes the protection provided by Apple useless as it will switch from r-x to rw- just fine when fault will happen

I hope I'm wrong about intentions of this hunk.

Reply:
I've put it behind an ASSERT. That mechanism has proven helpful when debugging.

Reply:
I thought about this little
While it's good mechanism to have, it shouldn't be enabled by default in debug builds as it will change the behaviour of fast/slow debug builds versus Release build.
I think the best approach would be to put this mechanism under non-product -XX option, and have  it disabled by default.
So it can be enabled only on debug build with special -XX flag.